http-client-jdk/src/main/java/io/micronaut/http/client/jdk/HttpResponseAdapter.java

```diff
@@ -0,0 +1,114 @@
+import io.micronaut.core.convert.value.MutableConvertibleValuesMap;
+import io.micronaut.core.type.Argument;
+import io.micronaut.http.HttpHeaders;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.HttpStatus;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.codec.MediaTypeCodec;
+import io.micronaut.http.codec.MediaTypeCodecRegistry;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+/**
+ * Adapter from {@link java.net.http.HttpResponse} to {@link HttpResponse}.
+ * @author Sergio del Amo
+ * @since 4.0.0
+ * @param <O> Body Type
+ */
+@Internal
+@Experimental
+public class HttpResponseAdapter<O> implements HttpResponse<O> {
+
+    private final java.net.http.HttpResponse<byte[]> httpResponse;
+    @NonNull
+    private final Argument<O> bodyType;
+    private final ConversionService conversionService;
+    private final MutableConvertibleValues<Object> attributes = new MutableConvertibleValuesMap<>();
+
+    private final MediaTypeCodecRegistry mediaTypeCodecRegistry;
+
+    public HttpResponseAdapter(java.net.http.HttpResponse<byte[]> httpResponse,
+                               @NonNull Argument<O> bodyType,
+                               ConversionService conversionService,
+                               MediaTypeCodecRegistry mediaTypeCodecRegistry) {
+        this.httpResponse = httpResponse;
+        this.bodyType = bodyType;
+        this.conversionService = conversionService;
+        this.mediaTypeCodecRegistry = mediaTypeCodecRegistry;
+    }
+
+    @Override
+    public HttpStatus getStatus() {
+        return HttpStatus.valueOf(httpResponse.statusCode());
+    }
+
+    @Override
+    public int code() {
+        return httpResponse.statusCode();
+    }
+
+    @Override
+    public String reason() {
+        return getStatus().getReason();
+    }
+
+    @Override
+    public HttpHeaders getHeaders() {
+        return new HttpHeadersAdapter(httpResponse.headers(), conversionService);
+    }
+
+    @Override
+    public MutableConvertibleValues<Object> getAttributes() {
+        return attributes;
+    }
+
+    @Override
+    public Optional<O> getBody() {
+        return convertBytes(getContentType().orElse(null), httpResponse.body(), bodyType);
+    }
+
+    private <T> Optional convertBytes(@Nullable MediaType contentType, byte[] bytes, Argument<T> type) {
+        if (type != null && mediaTypeCodecRegistry != null && contentType != null) {
+            if (CharSequence.class.isAssignableFrom(type.getType())) {
+                Charset charset = contentType.getCharset().orElse(StandardCharsets.UTF_8);
+                return Optional.of(new String(bytes, charset));
+            } else if (type.getType() == byte[].class) {
+                return Optional.of(bytes);
+            } else {
+                Optional<MediaTypeCodec> foundCodec = mediaTypeCodecRegistry.findCodec(contentType);
```

Comment:
rewrite using `flatMap`

Reply:
Like so?  I couldn't use `flatMap` as we have the catch-all method at the end of the function https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/9fbb719240df57c9cc8f5f0b040aed0de8927d84